gcam_to_cerf/get_capacity_factor.py

```diff
@@ -0,0 +1,85 @@
+import gcamreader
+import numpy as np
+import pandas as pd
+from pathlib import Path
+from standardize_output_format import *
+import os
+
+
+def get_query_by_name(queries, name):
+    return next((x for x in queries if x.title == name), None)
+
+
+def get_capacity_factor(
+    path_to_gcam_database:str,
+    gcam_file_name:str,
+    gcam_scenario:str,
+    capacity_crosscheck:pd.DataFrame,
+    save_output=False,
+    gcam_query_name = "elec investment capacity factor",
+    path_to_query_file: str = './elec_queries.xml',
+    ):
+
+    # create a Path from str
+    db_path = Path(path_to_gcam_database)
+
+    # create connection to gcam db
+    conn = gcamreader.LocalDBConn(db_path, gcam_file_name)
+
+    # parse the queries file
+    queries = gcamreader.parse_batch_query(path_to_query_file)
+
+    # collect capacity factor
+    capacity_factor = conn.runQuery(get_query_by_name(queries, gcam_query_name))
+
+    # rename columns
+    capacity_factor.rename(columns={
+        'Units': 'units',
+        'Year': 'year',
+        'value': 'capacity_factor',
+        }, inplace=True)
+
+    # calculate the mean capacity factor across investment segments
```

Comment:
I know when we talked in the past you were planning on back calculating new capacity using these capacity factors.  But in the previous PR it seems you are just getting it directly from the EJ-capacity data.

So then, I'm wondering what will you be using this data for.  In particular, I'm worried that take a straight `mean` is meaningless.  Perhaps a weighted mean by the level of investment in each investment segment (need a separate query for this) to get the average _expected_ capacity factor?

Reply:
@pralitp Correct, we are not using this capacity factor to back out new capacity investments. CERF needs capacity factor as a standalone variable to do calculations regarding expected operating costs over the expected technology lifetime.

A weighted average capacity factor makes sense to me and I can use the additional query you guys gave me (`elec investment by investment segment (energy)`) with energy by investment segment to get at some of it. However, since not all technologies that are purchased as capacity are operated to get energy, not all technologies needed appear in the energy by investment segment query. For these remaining technologies, we can just take the average capacity factor across investment segments, unless you have another idea?

Reply:
Great!

> However, since not all technologies that are purchased as capacity are operated to get energy, not all technologies needed appear in the energy by investment segment query. For these remaining technologies, we can just take the average capacity factor across investment segments, unless you have another idea?

Hmm, this is kind of a tricky question.  Just because we have a coal IGCC w/CCS plant in the subpeak segment as a hypothetical, that doesn't mean the economics would ever work out that you actually build it expecting a 10% capacity factor.

In GCAM we would capture that when doing the logit sharing, so we would be inclined to just include the coal IGCC w/CCS there even though it is very unlikely.  But if we take the average across investment segments here then you are kind of assuming people would think this is just as probable and outcome that using coal IGCC w/CCS as a subpeak technology or base load.

But the question remains: what to do in these cases.  Maybe expert judgment?  Higher capital cost technologies will bias towards the lower segments (base, intermediate) and lower capital costs towards higher segments (peak, subpeak).

Reply:
We could fill in the missing ones with the average by major technology type? For example, if coal IGCC recirculating is missing for 2030 Kentucky, we fill this in with the average of all the weighted average capacity factors for coal IGCC across all states and years? This way we would at least be capturing some of the investment segment behavior trends for each technology.

Alternatively we develop some generalized capacity factor assumptions to apply e.g., Nuclear = 80%, Natural Gas CC = 40%.

Reply:
Ah, yes this would be great.  I thought you were saying there were technologies that were missing from all states and years.

So then yes, I would say fall back to averages across all states / years.  Then if we still have to data, we fall back to generalized capacity factor assumptions.